chain/catchup.go

```diff
@@ -0,0 +1,123 @@
+package chain
+
+import (
+	"context"
+	"time"
+
+	// :(
+	logging "github.com/ipfs/go-log"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var logCatchup = logging.Logger("chain/catchup")
+
+type retryableError struct {
+	Err error
+}
+
+func (re retryableError) Error() string { return re.Err.Error() }
+
+func isRetryError(err error) bool {
+	if err == nil {
+		return false
+	}
+	_, ok := err.(retryableError)
+	return ok
+}
+
+type catchupStore interface {
+	GetHead() types.TipSetKey
+	GetTipSet(tsKey types.TipSetKey) (types.TipSet, error)
+}
+
+type catchupTracker interface {
+	SelectHead() (*types.ChainInfo, error)
+	UpdateTrusted(context.Context) error
+}
+
+type catchupSyncer interface {
+	HandleNewTipSet(context.Context, *types.ChainInfo, bool) error
+	Status() Status
+}
+
+type catchupComplete func(newHead *types.ChainInfo, curHead types.TipSet) (bool, error)
+
+type CatcherUpper struct {
+	tracker catchupTracker
+	syncer  catchupSyncer
+	store   catchupStore
+	done    catchupComplete
+}
+
+// NewCatcherUpper is a poorly named method that returns a CatcherUpper.
+func NewCatcherUpper(t catchupTracker, sy catchupSyncer, st catchupStore, done catchupComplete) *CatcherUpper {
+	return &CatcherUpper{
+		tracker: t,
+		syncer:  sy,
+		store:   st,
+		done:    done,
+	}
+}
+
+// Start starts a go routine to performe the catchup process.
+func (cu *CatcherUpper) Start(ctx context.Context, tick clockwork.Ticker) <-chan error {
```

Comment:
Accepting a ticker with the a goal of making testing easier using the new clock we have added. Using the ticker we should be able to control when this loop executes.